### PR TITLE
KAFKA-17032: NioEchoServer should generate meaningful id instead of incremental number.

### PR DESCRIPTION
**Pull request by frankvicky**
To aid in debugging, it's essential to generate meaningful IDs instead of using incremental numbers. This PR ensures that both the test and production environments share the same ID generation flow.
refer: #16384 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/network/ConnectionIdHelper.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.network;
+
+import java.net.Socket;
+
+public class ConnectionIdHelper {
+
+    public static String generateConnectionId(Socket socket, int nextConnectionIndex) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/network/ConnectionIdHelper.java]: Maybe we can move it into `Utils`?

File: clients/src/test/java/org/apache/kafka/common/network/NioEchoServer.java [hunk 2]
```diff
@@ -279,8 +278,10 @@ private static boolean maybeBeginServerReauthentication(KafkaChannel channel, Ne
         return false;
     }
 
-    private String id() {
-        return "connection-" + idGenerator.getAndIncrement();
+    private String id(SocketChannel channel) {
+        String connectionId = ConnectionIdHelper.generateConnectionId(channel.socket(), nextConnectionIndex.getAndIncrement());
+        nextConnectionIndex.compareAndSet(Integer.MAX_VALUE, 0);
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/network/NioEchoServer.java]: this protection may be a bit verbose since it is used by testing only. Also, we can inline it

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch.

File: clients/src/main/java/org/apache/kafka/common/network/ConnectionIdHelper.java (hunk 1 above)
Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/network/ConnectionIdHelper.java]: Cool, I will remove helper class.

**Comment by frankvicky** [comment]
Hi @chia7712 I have make some changes based on feedback, PTAL 😄

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 3]
```diff
@@ -1665,6 +1666,22 @@ public static ConfigDef mergeConfigs(List<ConfigDef> configDefs) {
...
     }
+
+    /**
+     * Generates a unique connection ID for the given socket.
+     *
+     * @param socket The socket for which the connection ID is to be generated.
+     * @param nextConnectionIndex The index to be used in the connection ID to ensure uniqueness.
+     * @return A string representing the unique connection ID.
+     */
+    public static String generateConnectionId(Socket socket, int nextConnectionIndex) {
```
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: nit: in this context, it can just be connectionIndex

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 4]
```diff
@@ -36,6 +36,7 @@
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Modifier;
+import java.net.Socket;
```
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Maybe this belongs as a utility method in Selector, since it's specific to the network subsystem, and this ID is meant to be used in Selector.

File: clients/src/test/java/org/apache/kafka/common/network/NioEchoServer.java [hunk 5]
```diff
@@ -90,7 +90,7 @@ public String metricNameSuffix() {
     private volatile boolean closeKafkaChannels;
     private final DelegationTokenCache tokenCache;
     private final Time time;
-    private final AtomicLong idGenerator = new AtomicLong();
+    private final AtomicInteger nextConnectionIndex = new AtomicInteger();
```
Comment (gharris1727) [inline clients/src/test/java/org/apache/kafka/common/network/NioEchoServer.java]: nit: this makes wraparound worse, can you apply the explicit wraparound logic that SocketServer uses?

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 4 above)
Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Yes, put it in the `Selector` is more reasonable. I will move it.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 3 above)
Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Yes, the method don't care about how to generate index.

File: clients/src/test/java/org/apache/kafka/common/network/NioEchoServer.java (hunk 5 above)
Comment (frankvicky) [reply clients/src/test/java/org/apache/kafka/common/network/NioEchoServer.java]: Sure, I will make sure they have same logic.

**Comment by frankvicky** [comment]
Hi @gharris1727, I have make some changes based on feedbacks, PTAL 😺

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 6]
```diff
@@ -1665,6 +1665,7 @@ public static ConfigDef mergeConfigs(List<ConfigDef> configDefs) {
         configDefs.forEach(configDef -> configDef.configKeys().values().forEach(all::define));
         return all;
     }
+
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: please revert this change

Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Sure.

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by chia7712** [comment]
@frankvicky could you please rebase code?

**Comment by frankvicky** [comment]
Hi @chia7712 , I have merged the latest trunk, PTAL

**Comment by chia7712** [comment]
@gharris1727 any feedback?

**Review by gharris1727 — APPROVED** [review]
LGTM, thanks @frankvicky for the PR!
